# Question 1

In [2]:
import pandas as pd

# feel free to just use the following if you prefer...
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

# Question 2

In [3]:
rows, columns = df.shape

print(f"The dataset has {rows} rows and {columns} columns.")

The dataset has 391 rows and 11 columns.


In [4]:
df.columns #chat bot says to print(df.columns) but print() not needed for jupyter notebooks

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

**Definitions**

1. Observations
--------------------------------------------------------------------------------------------------------------------------------
    a. Generally an observation or record is an entry or row in data sets which contain multiple values for
    variables listed in the columns.
    
        1. An attribute of observations are they can have values for each variable across the row
        2. Another attribute is observations are unique as they each represent a different entry for your data set
        
    b. In context to the csv file, an observation or entry would be the row of details for a villager which includes 
    the name, gender, birthday, etc.
--------------------------------------------------------------------------------------------------------------------------------
2. Variables
--------------------------------------------------------------------------------------------------------------------------------
    a. Variables in datasets often include the "details" of an observation. They often are defined in the
    columns of the data set and (hopefully) contain values for each entry.
    
        1. An attribute of variables is that they can hold numerous types of data such as numbers or 
        categories and even dates
        2. Another attribute is they often show a pattern or variance in a data.
    
    b. In context to the csv file, a variable can be the birthday or the name or gender of a villager.

# Question 3

In [5]:
#.describe() working examples (do i need the comments from ai?)
df.describe(include='all')  # To include both numeric and non-numeric columns

row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391               391   
unique            92      388               391   
top     K.K. Country  wee one  villager-admiral   
freq              10        2                 1   
mean             NaN      NaN               NaN   
std              NaN      NaN               NaN   
min              NaN      NaN               NaN   
25%              NaN      NaN               NaN   
50%              NaN      NaN               NaN   
75%              NaN      NaN               NaN   
max              NaN      NaN               NaN   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [6]:
df['species'].describe()  # To get a summary for just the 'species' column

count     391
unique     35
top       cat
freq       23
Name: species, dtype: object

In [7]:
#.value_counts() working example
df['species'].value_counts()  # Returns the count of each species in descending order

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [8]:
#another way of summarizing the code
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

In [9]:
df.tail()

row_n        id      name  gender  species birthday personality  \
386    475    winnie    Winnie  female    horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male     wolf    11-25      cranky   
388    480      yuka      Yuka  female    koala     7-20      snooty   
389    481      zell      Zell    male     deer      6-7        smug   
390    483    zucker    Zucker    male  octopus      3-8        lazy   

                song   phrase            full_id  \
386         My Place   hay-OK    villager-winnie   
387        K.K. Song  snarrrl  villager-wolfgang   
388     Soulful K.K.  tsk tsk      villager-yuka   
389         K.K. D&B    pronk      villager-zell   
390  Spring Blossoms    bloop    villager-zucker   

                                                   url  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...

# Question 4

New csv file from the question: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv

In [74]:
# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

In [75]:
df.shape

(891, 15)

In [76]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

a)
When you run .shape on the data set, it gives you the **total number** of rows and columns of the data set. If a data set had 8 rows and 7 columns, .shape would return (8,7). For .describe(), it will only (by default) return the columns that contain **numerical** values only. So if out of those 7 columns, only 3 had numerical values, it would only show those 3 columns.

b)
The discrepency in the count row, in column age, can be attributed to missing values. When we run .shape originally, it tells us that we have 891 rows for this data set. However, when we run the describe function we see that most row counts do indeed have 891 values, except for age. This tells us that age is missing 891-714 = 177 values or in simpler terms, 177 rows do not have a value for this variable.

# Question 5

The difference between an attribute and a method comes to what their purpose is and how they are called. The first distinction is an attribute does not need parentheses to be called, an example is .shape. While a method does need parantheses such as .describe(). The second distinction is what they return or their purpose. For the attribute it **provides** information about the data frame such as the number of columns or how many people values are in a row. A method usually has to **computate** values from the data. For example .describe() calculates the mean when called or the quartiles of the numerical data. A side note as well, when methods have empty parantheses, it usually means it has a default argument to work with. Once again for example, .describe()'s default value is all the columns with numerical values.

# Question 6

**Definitions:**

- count: This represents the number of all non-missing values in a column
- mean: This represents the average of the values of a column
- std: This represents the standard deviation of the values of a column
- min: This represents the smallest value in a column
- 25%: This represents the first quartile or Q1, where the first 25% of values of a column lay
- 50%: This represents the second quartile or Q2 or its proper name, the median, where the first 50% of values of a column lay
- 75%: This represents the third quartile or Q3, where the first 75% of values of a column lay
- max: This represents the largest value in a column 

Chatgpt summary for Q1-6: https://chatgpt.com/share/5b9b4993-4cd8-4dbd-a2e5-1d0d2028b75c

# Question 7

1. .dropna() might be more useful than deldf['col'] in cases where there are a small number of rows (relative to the data set) that are missing values.

2. deldf['col'] might be preferred because if we have a column that is missing maybe 60% or 70% or even 50% of its values, keeping such an empty column might negatively impact our analysis.

3. If we apply deldf['col'] before .dropna() there is a small chance that we can reduce the number of rows with empty values for variables without applying .dropna(), being able to keep more of the dataset. As well based off how many rows exist with empty values after deldf has been applied we can decide if we wish to keep these rows in our dataset for more variation.

In [77]:
#4.
#Before: url used is the: "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [78]:
#After:
del df['deck']
df.isna().sum()
df = df.dropna()
df.isna().sum()
#Now after removing one column and the age values we are left with no rows including 
#any missing values, making it a very full set

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

**Justification**
--------------------------------------------------------------------------------------------------------------------------------------
I first decided to remove the column name deck as it contained 688 missing values, meaning when I would drop the rows with the missing values I would only have to remove 181 (177 + 2 + 2) rows and not remove 688. This allowed the data set to have a large number of entries while having no missing values.

# Question 8

In [80]:
#1.
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

print(df.columns)

grouped_description = df.groupby('pclass')['age'].describe()

grouped_description

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [81]:
#2.
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#2 continued:

When you call .describe() it shows all the non-missing values in a column, for .groupby, the reason why the count is smaller is because they try to match each value in each column, and since age has a fewer amount of entries the overall dataframe will deal with fewer data points. 

In [1]:
#3 After Kernel has been reset

# #1
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

In [4]:
#Chatgpt's response:
import pandas as pd  # Import pandas

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

**Remarks for Error A**

I found it easier to ask chatgpt about my error compared to google. When I asked chatgpt, it would automatically give me the problem and offer a snippet of code that would fix the problem. When I would search it up on google, I would have to go through either chat forums or websites that had long documentation before I could figure out my reason.

In [6]:
#Error B, typo 1
url = 'titanics.csv'
df = pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [7]:
#Chatgpt's response:
#To fix the url or file name in my url variable 

# Load the Titanic dataset from your local file
#url = 'titanics.csv'  # Ensure the correct file name and path
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv' #I followed chatgpt's suggestion to fix the name
df = pd.read_csv(url)

# Check if the file loaded correctly
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [8]:
# typo 2

url = 'https://raw.githubusercontent.com/seaborn-data/master/titanics.csv' #titanic.csv -> titanics.csv & no mwaskom
df = pd.read_csv(url)

HTTPError: HTTP Error 400: Bad Request

In [9]:
#Chatgpt's response:

# Correct URL for the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

# Load the Titanic dataset from the correct URL
df = pd.read_csv(url)

# Display the first few rows to confirm
df.head()

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


**Remarks for Error B**

When it came to these two errors, I would say both google and chatgpt were very close in being the easiest to debug errors, but chatgpt wins again because of its ability to repair or suggest a concise fix compared to google searches bring long forums or complicated documentation. As well I felt I learned more using chatgpt as it would often guide me to the solution, allowing me to figure out the last step of the puzzle as with the first error (and even the second error if I had removed my comments before sharing it to chatgpt).

In [11]:
#3
DF.groupby("col1")["col2"].describe()

NameError: name 'DF' is not defined

In [12]:
#Chatgpt response:

# Define the DataFrame as df (lowercase)
df = pd.DataFrame({'col1': [1, 2, 3], 'col2': [4, 5, 6]})

df

col1  col2
0     1     4
1     2     5
2     3     6

**Remarks for C**

Once again chatgpt was far more efficient than google in finding the solution. Plus a pattern I started to learn was that if you use the same chat over and over again, the bot begins to learn what variables and syntax you are using and can curtail solutions to your exact needing making the coding process much more efficient. However, if you do not give it your exact code it often leaves it up to the coder to adjust their recommendations to your program.

In [13]:
#Error D

pd.read_csv(url

SyntaxError: incomplete input (1340441315.py, line 3)

In [14]:
#Chatgpt response:
#To check if I'm missing brackets or an incomplete statement or colons for breaks, etc.

#Fix:
pd.read_csv(url)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

**Remarks for Error D**

Even though chatgpt gave the right solution, I found its answer very verbose if the prompt is way too broad. I think if I kept the prompt far more narrow, maybe shared my own code with it then I would have gotten the solution faster. However, in this case because of chatgpt not being concise I found it quicker to find the problem with my code on google.

In [15]:
# Error E, typo 1
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.group_by("col1")["col2"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [18]:
#Chatgpt's Response:

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Group by 'pclass' and describe the 'age' column
grouped_description = df.groupby("pclass")["age"].describe()

grouped_description

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

In [19]:
#typo 2

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.groupby("col1")["col2"].describle()

KeyError: 'col1'

In [20]:
#Chatgpt's response:

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Group by 'pclass' and describe the 'age' column
grouped_description = df.groupby("pclass")["age"].describe()

grouped_description

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

**Remarks for Error E**

Similar to how I mentioned earlier, if you do not guide the chatbot with your code the solutions they offer can be correct but the coder has to adjust the solution to their code. However, despite this bump if you try again with your code, the solution is quickly given, making it more efficient than searching it up on google.

In [23]:
#Error F, typo 1

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

titanic_df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

In [24]:
#Chatgpt's response:

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

# Group by 'sex' and describe the 'age' column
grouped_description = titanic_df.groupby("sex")["age"].describe()

print(grouped_description)

        count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0


In [27]:
#typo 2

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

titanic_df.groupby("sex")["Age"].describe()

KeyError: 'Column not found: Age'

In [28]:
#Chatgpt's response: ---->> Note: I accidentally did the typo1 for Error G before Error F typo 2, so on the chatlog these 2 will be inverted

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

# Group by 'sex' and describe the 'age' column
grouped_description = titanic_df.groupby("sex")["age"].describe()

grouped_description

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

**Remarks for Error F**

What I found with using chatgpt in troubleshooting is that besides being able to get very concrete and concise debugging if you narrow the prompt well enough is besides it greater efficiency than google, it feels like your being taught by someone. Overtime with using it and using it I started to understand errors without needing to go to chatgpt, they became more recognizable and I became more comfortable with the language.

In [25]:
#Error G, typo 1

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

titanic_df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

In [26]:
#Chatgpt's response

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

# Group by 'sex' and describe the 'age' column
grouped_description = titanic_df.groupby("sex")["age"].describe()

grouped_description

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [29]:
#typo 2

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

titanic_df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

In [30]:
# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_df = pd.read_csv(url)

# Group by 'sex' and describe the 'age' column
grouped_description = titanic_df.groupby("sex")["age"].describe()

grouped_description

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

**Remarks for Error G**

Throughout question 8.3 I put chatgpt's response just to show how efficient it can be in terms of troubleshooting compared to google. When I used google I often found myself looking more for finding a similar person's experience online without getting help on the direction I needed to follow in terms of troubleshooting. I could have found a solution on google but figuring out how to implement it in my code proved to be very slow, with chatgpt on the other hand I found that its solutions either always pointed me in the right direction (if my prompt was too broad as I experienced earlier) or straight up figured out the solution if I made it narrow.

**Overall Conclusion:** I found chatgpt to be far more efficient in troubleshooting, learning, and familiarizing myself with the language, statistics, and problems.

Chatgpt summary for Q7-9: https://chatgpt.com/share/a60a5186-fd2c-495a-b65d-6e55d7be667e

# Question 9

Yes.